pkg/api/auth.go

```diff
@@ -13,19 +13,65 @@ import (
 	"time"
 
 	"github.com/anuvu/zot/errors"
+	"github.com/chartmuseum/auth"
 	"github.com/gorilla/mux"
 	"golang.org/x/crypto/bcrypt"
 )
 
-func authFail(w http.ResponseWriter, realm string, delay int) {
-	time.Sleep(time.Duration(delay) * time.Second)
-	w.Header().Set("WWW-Authenticate", realm)
-	w.Header().Set("Content-Type", "application/json")
-	WriteJSON(w, http.StatusUnauthorized, NewError(UNAUTHORIZED))
+const (
+	bearerAuthDefaultAccessEntryType = "repository"
+)
+
+func AuthHandler(c *Controller) mux.MiddlewareFunc {
+	if c.Config.HTTP.Auth != nil &&
+		c.Config.HTTP.Auth.Bearer != nil &&
+		c.Config.HTTP.Auth.Bearer.Cert != "" &&
+		c.Config.HTTP.Auth.Bearer.Realm != "" &&
+		c.Config.HTTP.Auth.Bearer.Service != "" {
+		return bearerAuthHandler(c)
+	}
+
+	return basicAuthHandler(c)
+}
+
+func bearerAuthHandler(c *Controller) mux.MiddlewareFunc {
+	authorizer, err := auth.NewAuthorizer(&auth.AuthorizerOptions{
+		Realm:           c.Config.HTTP.Auth.Bearer.Realm,
+		Service:         c.Config.HTTP.Auth.Bearer.Service,
+		PublicKeyPath:   c.Config.HTTP.Auth.Bearer.Cert,
+		AccessEntryType: bearerAuthDefaultAccessEntryType,
+	})
+	if err != nil {
+		panic(err)
```

Comment:
c.Log.Panic().Err(err).Msg("...")

Reply:
Noted and changed locally.  Will push an update once we have addressed the other comments raised